В цьому ДЗ потренуємось виконувати Z-тести для тестування гіпотез в різних задачах. Для коректної імплементації та інтерпретації тестів рекоемндую переглянути спершу документацію:

https://www.statsmodels.org/dev/generated/statsmodels.stats.weightstats.ztest.html

### Завдання 1: Z-тест для однієї вибірки

**Мета**: Перевірити, чи середнє значення вибірки відрізняється від гіпотетичного середнього.

**Припущення**: Середній ріст студентів університету вважається 170 см. Ми хочемо перевірити, чи середній ріст випадкової вибірки студентів істотно більший. Вибірка задана в змінній `heights`.

Для виконання завдання імпортуйте метод для проведення `ztest`, виконайте Z-тест з Python, виведіть p-value і зробіть заключення, чи справдилось припущення про середній зріст студентів на рівні значущості 0.05?

Використовуємо тут Z-тест, оскільки вибірка достатньо велика (більше 30 екземплярів) і можемо вважати, що її стандартне відхилення близьке до ст. відхилення генеральної сукупності, а значить - відоме.

In [1]:
import numpy as np
from statsmodels.stats.weightstats import ztest

In [2]:
heights = [174, 171, 175, 179, 170, 170, 179, 175, 169, 174, 169, 169, 173, 162, 163, 169, 166,
           173, 167, 164, 179, 170, 172, 164, 169, 175, 169, 174, 169, 169, 173, 162, 177]

In [33]:
null_mean = 170
alpha = 0.05

In [5]:
print("Середнє вибірки: {:.2f}".format(np.mean(heights)))
print("Стандартне відхилення вибірки: {:.2f}".format(np.std(heights)))

Середнє вибірки: 170.70
Стандартне відхилення вибірки: 4.65


In [7]:
ztest_Score, p_value = ztest(heights, value = null_mean, alternative='larger')

In [8]:
ztest_Score, p_value

(0.8482235371587619, 0.19815674571091746)

In [10]:
if(p_value <  alpha):
    print("Відхилити Н0.")
else:
    print("Н0 не може бути відхилена.")

Н0 не може бути відхилена.


**Висновки Завдання 1:**
* навіть просте обчислення середнього для вибірки вказує, що значення незначною мірою відрізняється від гіпотетичного середнього показника (170,7 см проти 170 см)
* Значення z-статистики незначне. Вона вказує, що різниця є несуттєвою.
* Нульова гіпотеза не може бути відхилена, оскільки p_value=0,20(заокруглив до 2 знаків після коми). Тобто середній зріст становить 170 см. 

### Завдання 2: Z-тест для двох незалежних вибірок

**Завдання 2.1.**

**Мета**: Перевірити, чи існує статистична різниця між середніми оцінками двох груп студентів.

**Припущення**: Група A отримала новий навчальний курс, а група B продовжувала зі стандартним курсом. Ми перевіряємо, чи є новий курс ефективнішим.

Виконайте Z-тест з Python, виведіть p-value і зробіть заключення, чи справдилось припущення щодо студентів на рівні значущості 0.05?

In [14]:
group_a_scores = [78.55, 72.25, 79.88, 75.  , 76.54, 74.99, 87.26, 77.93, 72.71,
       82.11, 71.9 , 79.04, 68.2 , 71.36, 78.98, 81.69, 78.86, 77.42,
       76.49, 70.61, 74.4 , 75.7 , 83.29, 79.72, 69.18, 79.62, 76.07,
       74.62, 81.06, 83.15, 82.66, 73.8 , 76.45, 79.66, 82.88, 75.6 ,
       77.07, 72.47, 72.02, 82.06]

group_b_scores = [81.78, 74.64, 80.02, 76.81, 71.77, 76.81, 82.69, 74.82, 82.82,
       61.9 , 79.11, 75.44, 73.5 , 75.46, 65.06, 73.9 , 76.79, 82.39,
       72.41, 70.96, 72.49, 79.58, 76.64, 72.35, 77.57, 75.49, 79.84,
       71.49, 73.36, 73.04, 67.68, 76.48, 76.31, 75.03, 73.83, 67.92,
       72.9 , 73.29, 70.99, 74.19]

In [15]:
np.mean(group_a_scores), np.mean(group_b_scores)

(77.08125, 74.73875000000001)

In [16]:
ztest(group_a_scores, group_b_scores, alternative='two-sided')

(2.357354864751502, 0.018405648552528006)

In [18]:
ztest(group_a_scores, group_b_scores, value=0) #щоб не забути другий варіант теж пробую

(2.357354864751502, 0.018405648552528006)

**Висновки Завдання 2.1:**
* Z-критерій становить 2.3573, а p_value = 0.0184.
* Оскільки p-value < 0.05, то ми маємо достатньо доказів, щоб відкинути нульову гіпотезу. Іншими словами, середній рівень оцінок відрізняється між двома класами.



**Завдання 2.2.**


**Мета**: Побачити, як впливає розмір вибірок на результат тесту.

**Завдання**: Уявіть, що з групи А ми маємо тільки перші 5 записів, а з групи В - усі. Таке могло статись, якщо ми вже маємо результати тестів учнів за попередньою програмою, а тести за новою програмою здали наразі лише 5 учнів і ми вирішили не чекати довше.
Виберіть перші 5 записів для групи А і проведіть t-test (для цього використовуємо наступний [метод](https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.ttest_ind.html)). Виведіть значення p-value. Зробіть висновок, який результат ви бачите щодо стат. значущості на рівні 0.05 для цього експерименту?

Тут проводимо t-test оскільки одна з вибірок - дуже мала.

In [23]:
group_a_scores_sample = group_a_scores[:5] #вибираю семпл з перших 5 учнів у класі А
group_a_scores_sample

[78.55, 72.25, 79.88, 75.0, 76.54]

In [26]:
from scipy.stats import ttest_ind #імпортую бібліотеку

In [27]:
ttest_ind(group_a_scores_sample, group_b_scores, equal_var=False)

TtestResult(statistic=1.1226527301193456, pvalue=0.3010736990344312, df=6.544857383782639)

**Висновки Завдання 2.2:**
* т-статистика 1.12 - є низькою, що свідчить про незначне відхилення середнього значення класу А від серед середнього в класі В
* p-value > 0.05 - це свідчить, що ми не можемо відхилити нульову гіпотезу. Різниця між двома група статистично незначуща.
* В підсумку, через невелику вибірку по класу А ми не змогли отримати адекватний результат, який ми вже знаємо на основі аналізу даних всього класу А. Мала вибірка негативно впливає і може приводити до помилкових висновків. Зокрема тут видається, що немає різниці між двома класами.


### Завдання 3\*: Z-тест для двох **пов'язаних** вибірок

**Мета**: Перевірити, чи вплинув тренінг на продуктивність працівників, коли всі справробітники проходили тренінг.

**Припущення**: Наші працівники проходять однаковий тренінг, і ми хочемо з'ясувати, чи покращилися їхні показники після тренінгу на рівні значущості 0.05. Виконайте Z-тест з Python, виведіть p-value і зробіть заключення, чи поліпшились показники працівників на рівні значущості 0.05?

Зверніть увагу, ці вибірки повʼязані між собою, значить не буде коректно виконати z-test між ними, натомість, ми можемо порівняти, чи різниця в результатах (продуктивності співробітників) є стат значущою.

In [29]:
before_training = [57.82, 37.63, 36.8 , 55.22, 52.97, 52.5 , 53.46, 43.2 , 52.32,
       52.93, 42.86, 68.66, 54.74, 38.09, 56.57, 40.25, 57.87, 61.59,
       41.79, 59.63, 54.13, 58.22, 68.97, 47.55, 42.46, 41.1 , 41.84,
       49.23, 53.41, 52.77]

after_training = [62.47, 40.66, 42.7 , 57.69, 61.41, 56.76, 54.75, 44.06, 56.29,
       55.48, 47.28, 72.6 , 57.59, 39.39, 56.54, 42.36, 62.58, 65.01,
       42.3 , 62.98, 57.9 , 59.45, 72.28, 50.66, 43.18, 44.82, 45.96,
       54.4 , 58.52, 53.01]

In [30]:
differences = np.array(after_training) - np.array(before_training)
print(differences)

[ 4.65  3.03  5.9   2.47  8.44  4.26  1.29  0.86  3.97  2.55  4.42  3.94
  2.85  1.3  -0.03  2.11  4.71  3.42  0.51  3.35  3.77  1.23  3.31  3.11
  0.72  3.72  4.12  5.17  5.11  0.24]


In [43]:
null_mean_2 = 0 #Наша нульова гіпотеза - різниця між резальтатами до та після тренінгу на рівні 0. Тобто тренінг нічого не дав. Тобто фактично альтернативна гіпотеза - різниця в результатах не дорівнює нулю.

In [44]:
ztest_Score, p_value = ztest(differences, value = null_mean_2, alternative='larger')
ztest_Score, p_value

(9.13890009623916, 3.1545359919175505e-20)

**Висновки Завдання 3:**
* z-статистика є дуже великою, що свідчить про суттєву різницю між середнім наших середніх показників порівняно з 0
* p_value дуууже малий, та відповідно < 0.05. Тобто ми відкидаємо нульову гіпотезу. Можливість випадково спостерігати такий результат майже дорівнює нулю. Тобто різниця в результах працівників є статистично значущою.